src/GafferScene/SetAlgo.cpp

```diff
@@ -0,0 +1,455 @@
+	}
+
+	void operator()( ExpressionAst const& ast )
+	{
+		boost::apply_visitor( *this, ast.expr );
+	}
+
+	void operator()( BinaryOp const& expr )
+	{
+		m_hash.append( expr.op );
+		boost::apply_visitor( *this, expr.left.expr );
+		boost::apply_visitor( *this, expr.right.expr );
+	}
+
+	const ScenePlug* m_scene;
+	IECore::MurmurHash &m_hash;
+
+};
+
+// Grammar
+// -------
+//  - Order of operator precedence: -, &, |
+//  - Space separated lists are supported (implicit OR)
+//  - Sets with a name that starts with / are NOT supported
+template <typename Iterator>
+struct ExpressionGrammar : qi::grammar<Iterator, ExpressionAst(), ascii::space_type>
+{
+	ExpressionGrammar() : ExpressionGrammar::base_type( expression )
+	{
+		using qi::_val;
+		using qi::_1;
+		using qi::char_;
+		using qi::lit;
+
+		expression =
+			orExpression [_val = _1];
+
+		orExpression =
+			  andExpression [_val = _1] >> *( '|' >> andExpression [_val /= _1] )
+			| element [_val = _1] >> *( '|' >> element [_val /= _1] );
+
+		andExpression =
+			  andNotExpression [_val = _1] >> *('&' >> andNotExpression [_val += _1])
+			| element [_val = _1] >> *('&' >> element [_val += _1]);
+
+		andNotExpression =
+			element [_val = _1] >> *( '-' >> element [_val -= _1] );
+
+		element =
+			  lit('(') >> orExpression [_val = _1] >> lit(')')
+			| lit('(') >> andExpression [_val = _1] >> lit(')')
+			| lit('(') >> andNotExpression [_val = _1] >> lit(')')
+			| lit('(') >> nameOrListOfNames [_val = _1] >> lit(')')
+			| nameOrListOfNames [_val = _1];
+
+		nameOrListOfNames =
+			  listOfNames [_val = _1]
+			| objectNameOrSetName [_val = _1];
+
+		listOfNames =
+			  lit('(') >> listOfNames [_val = _1] >> lit(')')
+			| objectNameOrSetName [_val = _1] >> *objectNameOrSetName [_val /= _1];
+
+		objectNameOrSetName =
+			  objectName [_val = _1]
+			| setName [_val = _1];
+
+		setName = char_( "a-zA-Z_" ) [ phx::at_c<0>(_val) = _1 ] >> *char_( "a-zA-Z_0-9" ) [ phx::at_c<0>(_val) += _1 ];
+
+		objectName = char_( "/" ) [ phx::at_c<0>(_val) = _1 ] >> *char_( "a-zA-Z_0-9/" ) [ phx::at_c<0>(_val) += _1 ];
+
+		// these have no effect unless BOOST_SPIRIT_DEBUG is defined
+		BOOST_SPIRIT_DEBUG_NODE(expression);
+		BOOST_SPIRIT_DEBUG_NODE(andNotExpression);
+		BOOST_SPIRIT_DEBUG_NODE(andExpression);
+		BOOST_SPIRIT_DEBUG_NODE(orExpression);
+		BOOST_SPIRIT_DEBUG_NODE(setName);
+		BOOST_SPIRIT_DEBUG_NODE(objectName);
+		BOOST_SPIRIT_DEBUG_NODE(nameOrListOfNames);
+		BOOST_SPIRIT_DEBUG_NODE(listOfNames);
+		BOOST_SPIRIT_DEBUG_NODE(objectNameOrSetName);
+	}
+
+	qi::rule<Iterator, SetName()> setName;
+	qi::rule<Iterator, ObjectName()> objectName;
+	qi::rule<Iterator, ExpressionAst(), ascii::space_type> expression, andNotExpression, andExpression, orExpression;
+	qi::rule<Iterator, ExpressionAst(), ascii::space_type> nameOrListOfNames, listOfNames, objectNameOrSetName, element;
+};
+
+void printSyntaxError( const std::string &setExpression, std::string::const_iterator position )
```

Comment:
Not sure if you like the way I handle the errors here, @johnhaddon?

```
WARNING : GafferScene::SetAlgo : Syntax error in indicated part of SetExpression.
setA - (/group/group/sphere2
     |---------------------|
```

Adding support for better errors would be a little tricky I guess. We could potentially run a check for unbalanced parantheses here, but meh.